### PR TITLE
[YUNIKORN-1337] Application state stuck in Accepted when placeholders are running and the job is deleted

### PR DESCRIPTION
**Pull request by pbacsko**
### What is this PR for?
YUNIKORN-1205 introduced a bug which causes an Application to stuck in "Accepted" state when the placeholders are running and the job is deleted.

### What type of PR is it?
* [x] - Bug Fix
* [ ] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-1337

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

**Review by manirajv06 — CHANGES_REQUESTED** [review]

File: pkg/scheduler/objects/application.go [hunk 1]
```diff
@@ -1502,13 +1502,17 @@ func (sa *Application) removeAllocationInternal(uuid string) *Allocation {
...
+	noAllocations := resources.IsZero(sa.pending) && resources.IsZero(sa.allocatedResource)
+
 	// update correct allocation tracker
 	if alloc.IsPlaceholder() {
 		sa.allocatedPlaceholder = resources.Sub(sa.allocatedPlaceholder, alloc.GetAllocatedResource())
 		// if all the placeholders are replaced, clear the placeholder timer
 		if resources.IsZero(sa.allocatedPlaceholder) {
 			sa.clearPlaceholderTimer()
-			if (sa.IsCompleting() && sa.stateTimer == nil) || sa.IsFailing() || sa.IsResuming() {
+			if (sa.IsCompleting() && sa.stateTimer == nil) || sa.IsFailing() || sa.IsResuming() || noAllocations {
```
[the remark this answers is not in the document]
Comment (pbacsko) [reply pkg/scheduler/objects/application.go]: Almost, but in the else block `allocatedResource` is updated, so it's not exactly the same.
Added a tiny function which increases readability.

File: pkg/scheduler/objects/application.go [hunk 2]
```diff
@@ -1531,7 +1528,7 @@ func (sa *Application) removeAllocationInternal(uuid string) *Allocation {
 	} else {
 		sa.allocatedResource = resources.Sub(sa.allocatedResource, alloc.GetAllocatedResource())
 		// When the resource trackers are zero we should not expect anything to come in later.
-		if resources.IsZero(sa.pending) && resources.IsZero(sa.allocatedResource) {
+		if sa.hasZeroAllocations() {
 			if err := sa.HandleApplicationEvent(CompleteApplication); err != nil {
```
Comment (manirajv06) [inline pkg/scheduler/objects/application.go:1532]: Can we make this as common block for both ph and normal alloc something like given below with var event being set with different values based on the various condition in the above code block (may be you can replace different if's with switch statement) ?

	// When the resource trackers are zero we should not expect anything to come in later.
	if sa.hasZeroAllocations() {
		event = CompleteApplication
	}

	if err := sa.HandleApplicationEvent(event); err != nil {
		log.Logger().Warn("Application state not changed while removing a placeholder allocation",
			zap.String("currentState", sa.CurrentState()),
			zap.String("event", event.String()),
			zap.Error(err))
	}

**Review by manirajv06 — APPROVED** [review]
looks good

**Comment by pbacsko** [comment]
codecov/patch can be ignored

**Pull request closed without merging**